### PR TITLE
Add interfaces for Multicall, Multicall2, and Multicall3

### PR DESCRIPTION
**Pull request by paulrberg**
As discussed in #53, providing interfaces is necessary so that this repo can be imported and used in a Foundry project.

New files:

- `interfaces/IMulticall.sol`
- `interfaces/IMulticall3.sol`
- `interfaces/IMulticall3.sol`

I have moved the structs in these files so that they are not duplicated, and have updated the tests to use the structs from here.

Note: I have formatted the interfaces with `forge fmt` using this [configuration](https://github.com/paulrberg/foundry-template/blob/c0d28ea3f46712ff5add6dff6aa27faeb0c0ed6d/foundry.toml#L3-L12).

### CODE REVIEW

File: src/Multicall.sol [hunk 1]
```diff
@@ -1,17 +1,13 @@
 // SPDX-License-Identifier: MIT
 pragma solidity 0.8.12;
 
+import {IMulticall} from "./interfaces/IMulticall.sol";
```
Comment (mds1) [inline src/Multicall.sol:4]: Thanks for this PR! I'd prefer to leave the Multicall contracts untouched and just duplicate the structs in the interface files. Although this isn't conventionally best practice, I want the `src` contracts on `master` to match what's already deployed/verified—and have future deployments match existing deployments—which means leaving the `src` files untouched. (The resulting bytecode _should_ be identical regardless, but verification would look different + solc can be unintuitive sometimes + metadata hash would change anyway)

Comment (paulrberg) [reply src/Multicall.sol]: Fair enough. I have pushed a new commit to undo the changes to the `src` contracts.

**Comment by paulrberg** [comment]
If this is merged, the commits should be squashed to avoid giving the impression in the git history that the `src` contracts have ever changed.

**Comment by mds1** [comment]
Good call, will definitely squash. I think previously you had updated tests to use e.g. `IMulticall3.Call3Value[]` instead of `Multicall3.Call3Value[]`, which seemed like a good sanity check on the interface correctness. Would you mind adding that back? If you prefer I'm happy to make those final changes before merging, since I appreciate all your recent contributions

**Comment by paulrberg** [comment]
Unfortunately, we cannot use the interface structs in the tests. I just tried and got this error:

```text
Compiler run failed
error[9553]: TypeError: Invalid type for argument in function call. Invalid implicit conversion from struct IMulticall.Call memory[] memory to struct Multicall.Call memory[] memory requested.
  --> src/test/Multicall.t.sol:25:76:
   |
25 |     (uint256 blockNumber, bytes[] memory returnData) = multicall.aggregate(calls);
   |
```

File: foundry.toml [hunk 2]
```diff
@@ -3,22 +3,22 @@
 # See more config options at: https://github.com/gakonst/foundry/tree/master/config
 
 # The Default Profile
-[default]
+[profile.default]
```
Comment (mds1) [inline foundry.toml:6]: Seems this PR needs a rebase, since these changes were already made in #54

File: src/interfaces/IMulticall.sol [hunk 3]
```diff
@@ -0,0 +1,27 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.5.0 <0.9.0;
```
Comment (mds1) [inline src/interfaces/IMulticall.sol:2]: If I try compiling this file with 0.5.0, 0.6.0 I get the below error:

```
IMulticallTest.sol:10:24: TypeError: Data location must be "calldata" for parameter in external function, but "memory" was given.
    function aggregate(Call[] memory calls)
                       ^-----------------^
```

So looks like it needs that change + `pragma experimental ABIEncoderV2;`

File: src/interfaces/IMulticall2.sol [hunk 4]
```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.5.0 <0.9.0;
```
Comment (mds1) [inline src/interfaces/IMulticall2.sol:2]: Same changes needed here

File: src/interfaces/IMulticall.sol (hunk 3 above)
Comment (paulrberg) [reply src/interfaces/IMulticall.sol:2]: Good catch! This repo seems like a good use case for [multibuild](https://github.com/foundry-rs/foundry/pull/2447). I wish I will found the time to add that feature to Foundry some day.

File: src/interfaces/IMulticall2.sol (hunk 4 above)
Comment (paulrberg) [reply src/interfaces/IMulticall2.sol:2]: Resolved in the commit below.

File: foundry.toml (hunk 2 above)
Comment (paulrberg) [reply foundry.toml:6]: Rebased!

File: src/interfaces/IMulticall.sol (hunk 3 above)
Comment (mds1) [reply src/interfaces/IMulticall.sol:2]: Awesome, thanks. Def would be useful, in the meantime setting up various CI build jobs [like this](https://github.com/foundry-rs/forge-std/blob/83c6e5b7bbbc865a83ed636ad32a8b7f2fc9185e/.github/workflows/ci.yml#L24-L60) is a decent workaround

**Review by mds1 — APPROVED** [review]
LGTM, thank you!

File: src/interfaces/IMulticall.sol (hunk 3 above)
Comment (paulrberg) [reply src/interfaces/IMulticall.sol:2]: I actually have started working on something similar - a GitHub Action for multibuild.

https://github.com/paulrberg/foundry-multibuild/blob/aa52a9c69840d3ca5af990b5a53d47c34ced8495/src/multibuild.sh

Comment (mds1) [reply src/interfaces/IMulticall.sol:2]: Looks like it might be private, getting a 404

Comment (paulrberg) [reply src/interfaces/IMulticall.sol:2]: Whoops, made it public now.

Comment (mds1) [reply src/interfaces/IMulticall.sol:2]: Ah nice, seems like it will be useful!

**Pull request merged**